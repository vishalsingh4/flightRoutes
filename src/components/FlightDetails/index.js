import { useEffect, useState } from 'react';
import clsx from 'clsx';

// mock-data import
import data from 'mock-data/flight-details.json';

// utils import
import {
    getFlightLogo,
    getDateFormat,
    getDifferenceInMinutes
} from 'utils';

// css imports
import './index.scss';

// constants
const timeFormat = 'HH:mm';
const dateFormat = 'EEE, dd LLL';

const airportCodeMap = {
    DEL: 'New Delhi International Airport',
    DXB: 'Dubai International Airport'
};

const FlightDetails = () => {
    const [flightData, setFlightData] = useState([]);
    const [selectedFlightId, setSelectedFlightId] = useState('');

    useEffect(() => {
        setFlightData(data);
    }, []);

    const handleFlightSelection = id => {
        setSelectedFlightId(id);
    };

    return (
        <div className="flight-details">
            <h3 className='search-heading color-red fs-20 fw-600 pt-4 m-0 text-center' tabIndex='0' aria-label='Search Flight Results'>Search Flight Results</h3>
            {
                flightData?.length > 0 && flightData.map(item => {
                    const {
                        id,
                        flightName,
                        flightNumber,
                        sourceCode,
                        arrivalTime,
                        destinationCode,
                        destinationTime,
                        price,
                        noOfStops
                    } = item;

                    const flightLogo = getFlightLogo(flightName);
                    const sourceName = airportCodeMap[sourceCode];
                    const destinationName = airportCodeMap[destinationCode];
                    const arrivalTimeInHM = getDateFormat(arrivalTime, timeFormat);
                    const arrivalTimeInDM = getDateFormat(arrivalTime, dateFormat);
                    const destinationTimeInHM = getDateFormat(destinationTime, timeFormat);
                    const destinationTimeInDM = getDateFormat(destinationTime, dateFormat);
                    const flightDuration = getDifferenceInMinutes(arrivalTime, destinationTime);
                    const isSelected = (selectedFlightId === id);
                    return (
                        <div className={clsx({
                            'flight-container container rounded d-flex align-items-center my-4 px-4 py-3 position-relative': true,
                            'border-red box-shadow': isSelected,
                            'border': !isSelected
                        })} key={id} tabIndex='0' aria-labelledby={`${flightName} flight from ${sourceName} to ${destinationName}; duration ${flightDuration}; price ${price}`}>
                            <section className='detail-section d-flex align-items-center'>
                                <div className='logo-section d-flex flex-column'>
                                    <img src={flightLogo} alt='flight-logo' width='50' height='50' />
                                    <span className='flight-name color-grey fs-15 fw-500'>{flightName}</span>
                                    <span className='flight-number color-grey fs-15 fw-500'>{flightNumber}</span>
                                </div>
                                <div className='source-section d-flex flex-column'>
                                    <span className='source-code color-dark fs-15 fw-bold'>{sourceCode}</span>
                                    <span className='source-time-hm color-dark fs-18 fw-bold'>{arrivalTimeInHM}</span>
                                    <span className='source-time-dm color-grey fs-15 fw-500'>{arrivalTimeInDM}</span>
                                    <span className='source-name color-grey fs-15 fw-500'>{sourceName}</span>
                                </div>
                                <div className="duration d-flex flex-column">
                                    <span className='flight-duration color-grey fs-15 fw-600'>{flightDuration}</span>
                                    <img src="https://fly10.ek.aero/Images/farebrand_refresh/flight@2x.png" alt="flight-duration" width='30' height='30' />
                                    <span className='noOf-stops color-grey fs-15 fw-600'>{noOfStops}</span>
                                </div>
                                <div className='destination-section d-flex flex-column'>
                                    <span className='destination-code color-dark fs-15 fw-bold'>{destinationCode}</span>
                                    <span className='destination-time-hm color-dark fs-18 fw-bold'>{destinationTimeInHM}</span>
                                    <span className='destination-time-dm color-grey fs-15 fw-600'>{destinationTimeInDM}</span>
                                    <span className='destination-name color-grey fs-15 fw-600'>{destinationName}</span>
                                </div>
                            </section>
                            <div className='vertical-separator' />
                            <section className='booking-section d-flex align-items-center'>
                                <span className='price-section color-red fs-20 fw-700 text-center'>&#x20b9; {price}</span>
                                <button className={clsx({
                                    'book-btn border-0 rounded h-45 fs-18 fw-700': true,
                                    'disabled color-dark': isSelected,
                                    'bg-red color-white': !isSelected
                                })} aria-label='Select Flight' onClick={() => handleFlightSelection(id)}>SELECT</button>
                            </section>
                        </div>
                    )
                })
            }
        </div >
    )
};

export default FlightDetails;