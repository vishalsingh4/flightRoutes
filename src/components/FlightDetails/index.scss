.flight-container {
    .detail-section {
        flex: 1 1 70%;
        .logo-section, .duration {
            flex: 1 0 20%;
        }
        .source-section, .destination-section {
            flex: 1 0 30%;
        }
    }

    .booking-section {
        flex: 1 0 30%;
        .price-section {
            flex: 1 0 40%;
        }
        .book-btn {
            flex: 1 0 60%;
        }
    }
    &:hover {
        box-shadow: rgba(0, 0, 0, 0.35) 0px 5px 15px;
    }
}

  /* Extra small devices (phones, 600px and down) */
  @media only screen and (max-width: 600px) {
    .main-section {
        top: 90px;
        max-height: calc(100% - 130px);
        .flight-details {
            padding-left: 20px;
            padding-right: 20px;
        }
        .flight-container {
            flex-direction: column;
            box-shadow: rgba(0, 0, 0, 0.35) 0px 5px 15px;
            .detail-section {
                flex-direction: column;
                flex: 1;
                text-align: center;
                .logo-section {
                    align-items: center;
                    .flight-name {
                        font-size: 12px;
                    }
                    .flight-number {
                        font-size: 12px;
                    }
                }
                .source-section {
                    padding: 10px;
                    .source-code {
                        font-size: 12px;
                    }
                    .source-time-hm {
                        font-size: 15px;
                    }
                    .source-time-dm {
                        font-size: 12px;
                    }
                    .source-name {
                        font-size: 12px;
                    }
                }
                .duration {
                    align-items: center;
                    .flight-duration {
                        font-size: 12px;
                    }
                    .noOf-stops {
                        font-size: 12px;
                    }
                }
                .destination-section {
                    padding: 10px;
                    .destination-code {
                        font-size: 12px;
                    }
                    .destination-time-hm {
                        font-size: 15px;
                    }
                    .destination-time-dm {
                        font-size: 12px;
                    }
                    .destination-name {
                        font-size: 12px;
                    }
                }
            }
            .vertical-separator {
                border-left: 0;
                border-bottom: 2px dashed #bcbcbc;
                width: 100%;
                height: 0;
                right: 0;
                top: 79%;
            }
            .booking-section {
                flex-direction: column;
                width: 100%;
                margin-top: 30px;
                .price-section {
                    font-size: 18px;
                    flex: 1;
                }
                .book-btn {
                    font-size: 15px;
                    flex: 1;
                    width: 75%;
                    padding: 6px;
                    margin-top: 10px;
                }
            }
        }
    }
  }

/* Small devices (portrait tablets and large phones, 600px and up) */
  @media only screen and (min-width: 600px) and (max-width: 768px) {
    .main-section {
        top: 104px;
        max-height: calc(100% - 157px);
        .flight-details {
            padding-left: 20px;
            padding-right: 20px;
        }
        .flight-container {
            flex-direction: column;
            box-shadow: rgba(0, 0, 0, 0.35) 0px 5px 15px;
            .detail-section {
                flex-direction: column;
                flex: 1;
                text-align: center;
                .logo-section {
                    align-items: center;
                    .flight-name {
                        font-size: 14px;
                    }
                    .flight-number {
                        font-size: 14px;
                    }
                }
                .source-section {
                    padding: 10px;
                    .source-code {
                        font-size: 14px;
                    }
                    .source-time-hm {
                        font-size: 18px;
                    }
                    .source-time-dm {
                        font-size: 14px;
                    }
                    .source-name {
                        font-size: 14px;
                    }
                }
                .duration {
                    align-items: center;
                    .flight-duration {
                        font-size: 14px;
                    }
                    .noOf-stops {
                        font-size: 14px;
                    }
                }
                .destination-section {
                    padding: 10px;
                    .destination-code {
                        font-size: 14px;
                    }
                    .destination-time-hm {
                        font-size: 18px;
                    }
                    .destination-time-dm {
                        font-size: 14px;
                    }
                    .destination-name {
                        font-size: 14px;
                    }
                }
            }
            .vertical-separator {
                border-left: 0;
                border-bottom: 2px dashed #bcbcbc;
                width: 100%;
                height: 0;
                right: 0;
                top: 79%;
            }
            .booking-section {
                flex-direction: column;
                width: 100%;
                margin-top: 30px;
                .price-section {
                    font-size: 20px;
                    flex: 1;
                }
                .book-btn {
                    font-size: 18px;
                    flex: 1;
                    width: 75%;
                    padding: 6px;
                    margin-top: 10px;
                }
            }
        }
    }
  }

/* Medium devices (landscape tablets, 768px and up) */
  @media only screen and (min-width: 768px) and (max-width: 992px) {
    .main-section {
        top: 125px;
        max-height: calc(100% - 200px);
        .flight-details {
            padding-left: 20px;
            padding-right: 20px;
        }
        .flight-container {
            flex-direction: column;
            box-shadow: rgba(0, 0, 0, 0.35) 0px 5px 15px;
            .detail-section {
                flex-direction: column;
                flex: 1;
                text-align: center;
                .logo-section {
                    align-items: center;
                    .flight-name {
                        font-size: 15px;
                    }
                    .flight-number {
                        font-size: 15px;
                    }
                }
                .source-section {
                    padding: 10px;
                    .source-code {
                        font-size: 15px;
                    }
                    .source-time-hm {
                        font-size: 20px;
                    }
                    .source-time-dm {
                        font-size: 15px;
                    }
                    .source-name {
                        font-size: 15px;
                    }
                }
                .duration {
                    align-items: center;
                    .flight-duration {
                        font-size: 15px;
                    }
                    .noOf-stops {
                        font-size: 15px;
                    }
                }
                .destination-section {
                    padding: 10px;
                    .destination-code {
                        font-size: 15px;
                    }
                    .destination-time-hm {
                        font-size: 20px;
                    }
                    .destination-time-dm {
                        font-size: 15px;
                    }
                    .destination-name {
                        font-size: 15px;
                    }
                }
            }
            .vertical-separator {
                border-left: 0;
                border-bottom: 2px dashed #bcbcbc;
                width: 100%;
                height: 0;
                right: 0;
                top: 79%;
            }
            .booking-section {
                flex-direction: column;
                width: 100%;
                margin-top: 30px;
                .price-section {
                    font-size: 22px;
                    flex: 1;
                }
                .book-btn {
                    font-size: 20px;
                    flex: 1;
                    width: 75%;
                    padding: 6px;
                    margin-top: 10px;
                }
            }
        }
    }
  } 

/* Large devices (laptops/desktops, 992px and up) */
  @media only screen and (min-width: 992px) and (max-width: 1200px){
    .main-section {
        .flight-details {
            .flight-container {
                .detail-section {
                    .logo-section, .duration {
                        flex: 1 1 15%;
                    }
                    .source-section,.destination-section {
                        flex: 1 1 35%;
                    }
                }
                .booking-section {
                    .price-section {
                        font-size: 18px;
                    }
                    .book-btn {
                        font-size: 15px;
                    }
                }
            }
        }
    }
  } 