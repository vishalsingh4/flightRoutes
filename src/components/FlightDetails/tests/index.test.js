import React from 'react';
import { shallow } from 'enzyme';
import FlightDetails from '..';

import data from 'mock-data/flight-details.json';

describe("FlightDetails Component", () => {
    const wrapper = shallow(<FlightDetails />);

    it('should render the FlightDetails Component correctly - snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });
    it('renders component', () => {
        expect(wrapper.exists()).toBe(true);
    });
    it('should render heading', () => {
        expect(wrapper.find('.search-heading').text()).toEqual('Search Flight Results');
    });
    it('should render mock data correctly', () => {
        const mockData = data;
        const mockDataLength = mockData && mockData.length;

        React.useState = jest.fn().mockReturnValue([mockData, () => { }]);

        const wrapper = shallow(<FlightDetails />);
        expect(wrapper.find('.flight-details').children().length).toEqual(mockDataLength + 1);
    });
    it('mock useEffect function', () => {
        const mockApiFn = jest.fn();
        React.useState = jest.fn().mockReturnValue([data, mockApiFn]);
        jest.spyOn(React, 'useEffect').mockImplementationOnce(f => f());

        shallow(<FlightDetails />);
        expect(mockApiFn).toHaveBeenCalled();
        expect(mockApiFn).toHaveBeenCalledTimes(1);
    });
    it('should select flight', () => {
        React.useState = jest.fn().mockReturnValueOnce([data, () => { }]).mockReturnValueOnce([1, () => { }]);

        const wrapper = shallow(<FlightDetails />);
        expect(wrapper.find('.border-red').exists()).toBe(true);
    });
    it('mock click function on select flight', () => {
        const mockTestData = [
            {
                id: 1,
                flightName: "EMIRATES",
                flightNumber: "EM1092",
                sourceCode: "DEL",
                arrivalTime: "30/09/2021 08:10:02",
                destinationCode: "DXB",
                destinationTime: "30/09/2021 14:10:02",
                price: "17424",
                noOfStops: "non-stop"
            }
        ];
        React.useState = jest.fn().mockReturnValueOnce([mockTestData, () => { }]).mockReturnValueOnce([1, () => { }]);

        const wrapper = shallow(<FlightDetails />);
        wrapper.find('.book-btn').simulate('click');
        expect(wrapper.find('.border-red').exists()).toBe(true);
    });
    it('should render test mock data correctly', () => {
        const mockTestData = [
            {
                id: 1,
                flightName: "EMIRATES",
                flightNumber: "EM1092",
                sourceCode: "DEL",
                arrivalTime: "30/09/2021 08:10:02",
                destinationCode: "DXB",
                destinationTime: "30/09/2021 14:10:02",
                price: "17424",
                noOfStops: "non-stop"
            }
        ];
        const mockData = mockTestData;
        const mockDataLength = mockData && mockData.length;

        React.useState = jest.fn().mockReturnValue([mockData, () => { }]);

        const wrapper = shallow(<FlightDetails />);
        expect(wrapper.find('.flight-details').children().length).toEqual(mockDataLength + 1);
        expect(wrapper.find('img').first().prop('src')).toEqual('emirates-tail.png');
        expect(wrapper.find('.flight-name').text()).toEqual(mockTestData[0].flightName);
        expect(wrapper.find('.flight-number').text()).toEqual(mockTestData[0].flightNumber);
        expect(wrapper.find('.source-code').text()).toEqual(mockTestData[0].sourceCode);
        expect(wrapper.find('.source-time-hm').text()).toEqual('08:10');
        expect(wrapper.find('.source-time-dm').text()).toEqual('Thu, 30 Sep');
        expect(wrapper.find('.source-name').text()).toEqual('New Delhi International Airport');
        expect(wrapper.find('.flight-duration').text()).toEqual('6hr 0mins');
        expect(wrapper.find('img').last().prop('src')).toEqual('https://fly10.ek.aero/Images/farebrand_refresh/flight@2x.png');
        expect(wrapper.find('.noOf-stops').text()).toEqual(mockTestData[0].noOfStops);
        expect(wrapper.find('.destination-code').text()).toEqual(mockTestData[0].destinationCode);
        expect(wrapper.find('.destination-time-hm').text()).toEqual('14:10');
        expect(wrapper.find('.destination-time-dm').text()).toEqual('Thu, 30 Sep');
        expect(wrapper.find('.destination-name').text()).toEqual('Dubai International Airport');
        expect(wrapper.find('.price-section').text()).toEqual(`â‚¹ ${mockTestData[0].price}`);
        expect(wrapper.find('.book-btn').text()).toEqual('SELECT');
    });
})