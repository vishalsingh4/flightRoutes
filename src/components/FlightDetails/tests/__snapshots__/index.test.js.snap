// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlightDetails Component should render the FlightDetails Component correctly - snapshot 1`] = `
<div
  className="flight-details"
>
  <h3
    aria-label="Search Flight Results"
    className="search-heading color-red fs-20 fw-600 pt-4 m-0 text-center"
    tabIndex="0"
  >
    Search Flight Results
  </h3>
</div>
`;
